kartothek/utils/pandas.py

```diff
@@ -0,0 +1,373 @@
+        # approach effectively reduces the maximum memory consumption and to our
+        # knowledge is not measuable slower.
+        colset = set(dfs[0].columns)
+        if not all(colset == set(df.columns) for df in dfs):
+            raise ValueError("Not all DataFrames have the same set of columns!")
+
+        res = pd.DataFrame(index=pd.RangeIndex(sum(len(df) for df in dfs)))
+        for col in dfs[0].columns:
+            res[col] = pd.concat(
+                [df[col] for df in dfs], ignore_index=True, sort=False, copy=False
+            )
+
+    # ensure list (which is still referenced in parent scope) gets emptied
+    del dfs[:]
+
+    return res
+
+
+def is_dataframe_sorted(df, columns):
+    """
+    Check that the given DataFrame is sorted as specified.
+
+    This is more efficient than sorting the DataFrame.
+
+    An empty DataFrame (no rows) is considered to be sorted.
+
+    .. warning::
+
+        This function does NOT handle NULL values correctly!
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        DataFrame to check.
+    colums: Iterable[str]
+        Column that the DataFrame should be sorted by.
+
+    Returns
+    -------
+    sorted: bool
+        ``True`` if DataFrame is sorted, ``False`` otherwise.
+
+    Raises
+    ------
+    ValueError: If ``columns`` is empty.
+    KeyError: If specified columns in ``by`` is missing.
+    """
+    columns = list(columns)
+
+    if len(columns) == 0:
+        raise ValueError("`columns` must contain at least 1 column")
+
+    state = None
+
+    for col in columns[::-1]:
+        data = df[col].values
+        if isinstance(data, pd.Categorical):
+            data = np.asarray(data)
+        data0 = data[:-1]
+        data1 = data[1:]
+
+        with np.errstate(invalid="ignore"):
+            comp_le = data0 < data1
+            comp_eq = data0 == data1
+
+        if state is None:
+            # last column
+            state = comp_le | comp_eq
+        else:
+            state = comp_le | (comp_eq & state)
+
+    return state.all()
+
+
+def sort_dataframe(df, columns):
+    """
+    Sort DataFrame by columns.
+
+    This is roughly equivalent to::
+
+        df.sort_values(columns).reset_index(drop=True)
+
+    .. warning::
+
+        This function does NOT handle NULL values correctly!
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        DataFrame to sort.
+    columns: Iterable[str]
+        Columns to sort by.
+
+    Returns
+    -------
+    df: pandas.DataFrame
+        Sorted DataFrame w/ reseted index.
+    """
+    columns = list(columns)
+    if is_dataframe_sorted(df, columns):
+        return df
+    data = [df[col].values for col in columns[::-1]]
+    df = df.iloc[np.lexsort(data)]
+    # reset inplace to reduce the memory usage
+    df.reset_index(drop=True, inplace=True)
+    return df
+
+
+def mask_sorted_duplicates_keep_last(df, columns):
```

Comment:
is this still used? Sounds like being connected to the timestamp feature we dropped

Reply:
It is still used in https://github.com/JDASoftwareGroup/kartothek/pull/307/files#diff-d436e798a976e33049d773d78d59a747R234-R236 .
If this functionality has to be removed, related code cleanup will be done.
@marco-neumann-jdas , please comment on the same.

Reply:
It is used, since in theory it is possible that the datasets where written with non-unique key entries (aka duplicates) and the user always gets a consistent, non-broken result. I'm OK with changing that to a check with a hard error in the future (not this PR).